setup.py

```diff
@@ -196,6 +210,33 @@ def run(self):
         # It's an old-style class in Python 2.7...
         distutils.command.clean.clean.run(self)
 
+def throw_of_failure(command):
+    ret = os.system(command)
+    if ret != 0:
+        raise Exception("{} failed".format(command))
+
+
+def build_deps():
+    this_dir = os.path.dirname(os.path.abspath(__file__))
+    if sys.platform.startswith('linux'):
+        cpu_count = multiprocessing.cpu_count()
+        os.chdir("third_party/zlib/")
+        throw_of_failure('cmake .')
+        throw_of_failure("cmake --build . --parallel {}".format(cpu_count))
+        os.chdir(this_dir)
+
+        zlib_path = os.path.join(this_dir, "third_party/zlib")
+        libpng_cmake_options="-DPNG_BUILD_ZLIB=ON -DPNG_STATIC=OFF -DZLIB_INCLUDE_DIR:PATH={zlib_path} -DZLIB_LIBRARY:FILEPATH={zlib_path}/libz.so".format(zlib_path=zlib_path)
```

Comment:
Try setting "PNG_STATIC=ON". Also see [https://stackoverflow.com/questions/43772887/cmake-creating-a-static-library](https://stackoverflow.com/questions/43772887/cmake-creating-a-static-library) or related articles.

Reply:
Hi, 

I Already tried that. That was actually my preferred option.  I spent a lot of time trying to link statically without success. Whatever the combination of options I tried, linking always fails with obscure errors.  The problem is that to link statically, we would need to compile position independent code ( with the -fPIC option) which for these libs does not seem to be working correctly.  

I am close to figuring out how to ship the dynamic libraries with the package. I'll keep you posted.

Reply:
This is [an example of a statically built library](https://github.com/pytorch/pytorch/blob/bc2e6d10fa7dc73b989ab3e22126867b50fed39c/aten/src/ATen/cpu/tbb/CMakeLists.txt) in PyTorch. You can use the [LD_DEBUG](http://man7.org/linux/man-pages/man8/ld.so.8.html) environment variable to print more information and see what libraries you're linking against so you can confirm the system shared libraries aren't being used. You can also use [the version macros](https://github.com/glennrp/libpng/blob/216387f6f7ee043df24a84ab2dc53d2827148b5e/CMakeLists.txt#L27) for the built libraries as another point of verification, assuming they're differing from the system library (I'd assume that the system libraries are older).

Another thing to add on top of this is uninstalling the linked libraries from the CI, but since they're so essential this might cause all kinds of unintended removals.

Reply:
We actually don't ship dynamic libraries anymore, because they can lead to all kinds of nightmares / conflicts with a given system.

Reply:
Could you please elaborate on the dynamic libraries part ? 
I just ran `ldd` on the `_C.so` extension for the torch nightly with the following output on my virtualenv.
```
	linux-vdso.so.1 (0x00007ffe4f261000)
	libshm.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libshm.so (0x00007ff98e5f9000)
	libtorch_python.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libtorch_python.so (0x00007ff986e26000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007ff986a9d000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ff9866ff000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007ff9864e7000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007ff9862c8000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ff985ed7000)
	libtorch.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libtorch.so (0x00007ff985cc3000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007ff985abb000)
	libtorch_cpu.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libtorch_cpu.so (0x00007ff976a5b000)
	libtorch_cuda.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libtorch_cuda.so (0x00007ff937be5000)
	libnvToolsExt-3965bdd0.so.1 => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libnvToolsExt-3965bdd0.so.1 (0x00007ff9379db000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007ff9377d7000)
	libc10_cuda.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libc10_cuda.so (0x00007ff9375a8000)
	libc10.so => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libc10.so (0x00007ff93734a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007ff98ea1e000)
	libgomp-7c85b1e2.so.1 => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libgomp-7c85b1e2.so.1 (0x00007ff937120000)
	libcudart-1b201d85.so.10.1 => /home/ryad/.pyvenvs/torchvision/lib/python3.7/site-packages/torch/./lib/libcudart-1b201d85.so.10.1 (0x00007ff936ea1000)
```
Here is what the `lib` folder in the `torch` nightly package contains : 
```  
libc10.so       
libcaffe2_detectron_ops_gpu.so    
libcaffe2_nvrtc.so      
libcudart-1b201d85.so.10.1  
libnvrtc-5e8a26c9.so.10.1  
libnvToolsExt-3965bdd0.so.1  libtorch.so      
libtorch_cuda.so         
libtorch_python.so
libc10_cuda.so  
libcaffe2_module_test_dynamic.so  
libcaffe2_observers.so  
libgomp-7c85b1e2.so.1       
libnvrtc-builtins.so       
libshm.so                    
libtorch_cpu.so  
libtorch_global_deps.so 
```
These are all dynamic libraries shipped with `torch`. 

I have just figured out how to do the same while working on the `jpeg` library. I intend to do this here. Would you be okay with it ?